src/software/geom/spline.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <vector>
+
+#include "software/geom/point.h"
+#include "software/geom/polynomial.h"
+
+/**
+ * A Spline maps a double to a Point with a parameterized piecewise function
+ * composed of Polynomials over ranges
+ */
+class Spline
+{
+   public:
+    Spline() = delete;
+    /**
+     * Construct a spline by drawing line segments between consecutive
+     * Points
+     *
+     * @throws std::invalid_argument if points.size() == 0
+     *
+     * @param points Points on the spline
+     */
+    explicit Spline(const std::vector<Point>& points);
+
+    /**
+     * Construct a spline by drawing line segments between consecutive
+     * Points
+     *
+     * @throws std::invalid_argument if points.size() == 0
+     *
+     * @param points Points on the spline
+     */
+    Spline(const std::initializer_list<Point>& points);
+
+    /**
+     * Calculates the value of spline evaluated at value val
+     *
+     * @param val value to evaluate spline
+     *
+     * @throws std::invalid_argument if val is outside of domain
+     *
+     * @return value of spline evaluated at value val
+     *      * if not defined by a spline then return closest start/end point
+     */
+    Point valueAt(double val) const;
+
+    /**
+     * Gets the number of knots in the spline including start and end points
+     *
+     * @return size of the spline
+     */
+    size_t size(void) const;
+
+    /**
+     * Gets knots in the spline including start and end points
+     *
+     * @return knots in the spline
+     */
+    const std::vector<Point> getKnots(void) const;
+
+    /**
+     * Gets domain of the spline
+     *
+     * @return domain of the spline
+     */
+    const std::pair<double, double> getDomain(void) const;
+
+   private:
+    class SplineSegment
+    {
+       public:
+        SplineSegment(Polynomial x, Polynomial y, double start, double end)
+            : x(x), y(y), start(start), end(end)
+        {
+        }
+        const Polynomial x;
+        const Polynomial y;
+        const double start;
+        const double end;
+    };
+
+    // segments represent the polynomials that interpolate between
+    // time at index to index + 1
+    std::vector<SplineSegment> segments;
+
+    // points that connect segments
+    const std::vector<Point> knots;
+
+    // the interval over which the spline has defined points
+    const std::pair<double, double> domain;
```

Comment:
Since we're storing the domain like this, does this mean we're supporting splines that are not in `[0,1]`?

Reply:
I guess we're not... removed